README.md
<!-- Improved compatibility of back to top link: See: https://github.com/othneildrew/Best-README-Template/pull/73 -->
<a name="readme-top"></a>
<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
*** or simply open an issue with the tag "enhancement".
*** Don't forget to give the project a star!
*** Thanks again! Now go create something AMAZING! :D
-->



<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![Commitizen friendly][commitizen-shield]][repository-url]
[![npm monthly downloads][npm-downloads-shield]][npm-downloads-url]
[![current version][current-version-shield]][repository-url]
[![LinkedIn][linkedin-shield]][linkedin-url]



<!-- PROJECT LOGO -->
<br />
<div align="center">

<h3 align="center">TypeScript toolbox</h3>

  <p align="center">
    A collection of TypeScript utilities
    <br />
    <a href="https://lerado.github.io/typescript-toolbox"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/Lerado/typescript-toolbox/issues">Report Bug</a>
    ·
    <a href="https://github.com/Lerado/typescript-toolbox/.github/issues">Request Feature</a>
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

This library is a comprehensive collection of TypeScript utilities that can be used to simplify and streamline TypeScript projects. It includes a wide range of functions, classes, and interfaces that make it easier to work with common data types, manipulate arrays and objects, and manage other aspects of the language. Additionally, the library is designed to be easy to use and integrate into existing projects, making it a valuable resource for TypeScript developers of all levels.

<p align="right">(<a href="#readme-top">back to top</a>)</p>



### Built With

* [![TypeScript][TypeScript]][Typescript-url]

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- GETTING STARTED -->
## Getting Started

To start using this library in your project, follow these simple example steps.

### Prerequisites

You basically just need `npm` or `yarn`:
* npm
  ```sh
  npm install npm@latest -g
  ```

* yarn
  ```sh
  npm install yarn@latest -g
  ```

### Installation

Just a single command required. Depending on which package manager you use.
* npm
  ```sh
  npm install @lerado/typescript-toolbox
  ```

* yarn
  ```sh
  yarn add @lerado/typescript-toolbox
  ```


<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- USAGE EXAMPLES -->
## Usage

Import an use any function, decorator, etc...

```ts
import { objectToFlatString } from '@lerado/typescript-toolbox';

objectToFlatString({ id: 123, name: 'Mr. Smith', age: 23 });
// => to 123mr. smith23.
```

_Please refer to the [documentation](https://lerado.github.io/typescript-toolbox) for more._

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- ROADMAP -->
## Roadmap

- [x] Release v1.0.0 on npm
- [ ] Configure Github actions for CI/CD
<!-- - [ ] Feature 2
- [ ] Feature 3
    - [ ] Nested Feature -->

See the [open issues](https://github.com/Lerado/typescript-toolbox/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

We use the [Git workflow](https://git-flow.readthedocs.io/fr/latest/presentation.html) and [Commitizen](https://github.com/commitizen) for our development cycles and releases. We **strongly recommend** you use it too with your contributions.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your feature/hotfix Branch (`git checkout -b (feature|hotfix)/branch-name`)
3. Commit your Changes (`git commit -m <COMMITIZEN FRIENDLY MESSAGE>`)
4. Push to the Branch (`git push origin <YOUR BRANCH>`)
5. Open a Pull Request

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Darel Ngogang - [@JeSuisLerado](https://twitter.com/@JeSuisLerado) - darel.ngogang@gmail.com

Project Link: [https://github.com/Lerado/typescript-toolbox](https://github.com/Lerado/typescript-toolbox)

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* [TypeScript Starter template](https://github.com/bitjson/typescript-starter)
* [How to create and publish a TypeScript library](https://medium.com/counterarts/a-beginners-guide-to-creating-a-typescript-library-and-a-github-repository-26ae3f1540ad)

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[current-version-shield]: https://img.shields.io/npm/v/@lerado/typescript-toolbox.svg?style=for-the-badge
[commitizen-shield]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?style=for-the-badge
[npm-downloads-shield]: https://img.shields.io/npm/dm/@lerado/typescript-toolbox.svg?style=for-the-badge
[npm-downloads-url]: https://www.npmjs.com/package/@lerado/typescript-toolbox
[contributors-shield]: https://img.shields.io/github/contributors/Lerado/typescript-toolbox.svg?style=for-the-badge
[contributors-url]: https://github.com/Lerado/typescript-toolbox/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/Lerado/typescript-toolbox.svg?style=for-the-badge
[forks-url]: https://github.com/Lerado/typescript-toolbox/network/members
[stars-shield]: https://img.shields.io/github/stars/Lerado/typescript-toolbox.svg?style=for-the-badge
[stars-url]: https://github.com/Lerado/typescript-toolbox/stargazers
[issues-shield]: https://img.shields.io/github/issues/Lerado/typescript-toolbox.svg?style=for-the-badge
[issues-url]: https://github.com/Lerado/typescript-toolbox/issues
[license-shield]: https://img.shields.io/github/license/Lerado/typescript-toolbox.svg?style=for-the-badge
[license-url]: https://github.com/Lerado/typescript-toolbox/blob/master/LICENSE
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/darel-ngogang-00b8a6223
[TypeScript]: https://img.shields.io/badge/typescript-000000?style=for-the-badge&logo=typescript
[Typescript-url]: https://www.typescriptlang.org/
[repository-url]: https://github.com/Lerado/typescript-toolbox


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/Lerado/typescript-toolbox/compare/v1.0.1...v1.1.0) (2023-06-15)


### Features

* add breadthFirstTraversal function in utils ([184a2b4](https://github.com/Lerado/typescript-toolbox/commit/184a2b4f13302d8f2ce63d5fa91a4d9dc04a85db))

### [1.0.1](https://github.com/Lerado/typescript-toolbox/compare/v1.0.0...v1.0.1) (2023-06-15)


### Bug Fixes

* export utils from library index ([7905d81](https://github.com/Lerado/typescript-toolbox/commit/7905d813a1d9a2a7188df71571c7604c751ee603))

## 1.0.0 (2023-06-02)


### Features

* add @OnChange, depthFirstTraversal, objectToFlatString and isStrongPassword ([21d7c5a](https://github.com/Lerado/typescript-toolbox/commit/21d7c5afdc5554857293c3b603928bed90a786fb))


src/lib/utils/strong-password.spec.ts
import test from 'ava';

import { isStrongPassword } from './strong-password';

test('isStrongPassword', (t) => {
  t.false(isStrongPassword('123456'));
  t.false(isStrongPassword('password123'));
  t.false(isStrongPassword('Password123'));
  t.true(isStrongPassword('Password!123'));
});


src/lib/utils/depth-first-traversal.spec.ts
import test from 'ava';

import { depthFirstTraversal } from './depth-first-traversal';

test('depthFirstTraversal has correct output', (t) => {
  const tree = {
    id: 1,
    children: [
      { id: 2, children: [{ id: 4 }] },
      { id: 3, children: [{ id: 5 }] },
    ],
  };
  t.log(depthFirstTraversal(tree, 'children').map((node) => node.id));
  t.deepEqual(
    depthFirstTraversal(tree, 'children').map((node) => node.id),
    [1, 2, 4, 3, 5]
  );
});

test('depthFirstTraversal callback', (t) => {
  const tree = {
    id: 1,
    children: [
      { id: 2, children: [{ id: 4 }] },
      { id: 3, children: [{ id: 5 }] },
    ],
  };
  let callbackExecutionCounter = 0;
  depthFirstTraversal(tree, 'children', () => callbackExecutionCounter++);
  // Callback should have executed 5 times
  t.is(callbackExecutionCounter, 5);
});


src/lib/utils/object-to-flat-string.spec.ts
import test from 'ava';

import { objectToFlatString } from './object-to-flat-string';

test('objectToFlatString', (t) => {
  t.is(
    objectToFlatString({
      id: 123,
      name: 'Mr. Smith',
      profile: { age: 23, linkedInUrl: null },
      deletedAt: null,
    }),
    '123mr. smith23'
  );
});


src/lib/utils/depth-first-traversal.ts
/**
 * Performs a depth-first traversal of a tree-like structure
 * and returns the march of the traversal
 *
 * ### Usage
 * ```ts
 * import { depthFirstTraversal } from '@lerado/typescript-toolbox';
 *
 * const tree = {
 *  id: 1,
 *  children: [
 *    { id: 2, children: [ { id: 4 } ] },
 *    { id: 3, children: [ { id: 5 } ] }
 *  ]
 * };
 *
 * depthFirstTraversal(tree, 'children', console.log);
 * // => console: 1, 2, 4, 3, 5
 * // => [{ id: 1 }, { id: 2 }, { id: 4 },  { id: 3 },  { id: 5 }]
 * ```
 * @param root - the tree's root object
 * @param childrenKey - accessor name for child elements
 * @param callback - a callback function to execute for each visited element
 */
const depthFirstTraversal = <T extends Record<string, unknown>>(
  root: T,
  childrenKey: string,
  callable?: (visitedElement: T) => unknown
): Partial<T>[] => {
  const result: T[] = [];
  const stack: T[] = [];
  stack.push(root);

  while (stack.length) {
    const visitedElement: T = stack.pop() as T;
    // Visit
    if (callable) {
      callable(visitedElement);
    }
    result.push(visitedElement);
    // Stack children in the reverse order to preserve original order
    const children: T[] = (visitedElement[childrenKey] as T[]) ?? [];
    for (let childIndex = children.length - 1; childIndex >= 0; childIndex--) {
      stack.push(children[childIndex]);
    }
  }

  return result;
};

export { depthFirstTraversal };


src/lib/utils/object-to-flat-string.ts
/**
 * Serializes an object.
 * The idea is to serialize each object to a queryable string.
 * This is very useful for search engines.
 *
 * ### Usage
 * ```ts
 * import { objectToFlatString } from '@lerado/typescript-toolbox';
 *
 * objectToFlatString({ id: 123, name: 'Mr. Smith', age: 23 });
 * // => to 123mr. smith23.
 * ```
 * @param object
 */
const objectToFlatString = (object: Record<string, unknown>): string => {
  // A treatment stack
  const stack = [];
  let result = '';

  // Initialization
  for (const key in object) {
    if (object[key] !== null) {
      stack.push(object[key]);
    }
  }
  stack.reverse(); // Just to pop in the same order of the keys

  // Iteration and serialization
  while (stack.length) {
    const value: unknown = stack.pop();

    if (typeof value === 'object') {
      for (const key in value) {
        const keyValue = (value as Record<string, number>)[key];
        if (keyValue !== null) {
          stack.push(keyValue);
        }
      }
    } else {
      result += String(value).toLowerCase();
    }
  }

  return result;
};

export { objectToFlatString };


src/lib/decorators/onchange.decorator.ts
export type SimpleChange<T> = {
  readonly firstChange: boolean;
  readonly previousValue: T;
  readonly currentValue: T;
} & { readonly isFirstChange: () => boolean };

type CallbackFunction<T> = (value: T, change: SimpleChange<T>) => unknown;

/**
 * Invokes a callback method when a class property is set
 *
 * ### Usage
 * ```ts
 * import { OnChange } from '@lerado/typescript-toolbox';
 *
 * class MyClass {
 *  &#64;OnChange(function(changes) {
 *    console.log(changes);
 *  })
 *  property1 = 0;
 *
 *  &#64;OnChange(function(changes) {
 *    console.log(changes);
 *  })
 *  property1 = 0;
 * }
 * ```
 * @param callback - the callback function
 */
const OnChange = <T = unknown>(
  callback: CallbackFunction<T> | string
): CallableFunction => {
  const cachedValueKey = Symbol();
  const isFirstChangeKey = Symbol();

  return (target: Record<string, unknown>, key: PropertyKey) => {
    const callbackFn: CallbackFunction<T> =
      typeof callback === 'string'
        ? (target[callback] as CallbackFunction<T>)
        : callback;
    if (!callbackFn) {
      return new Error(
        `Cannot find method ${callback} in class ${target.constructor.name}`
      );
    }

    return Object.defineProperty(target, key, {
      set: function (value) {
        // Change status of "isFirstChange"
        this[isFirstChangeKey] = this[isFirstChangeKey] === undefined;

        // Do nothing if old value is the same as new value
        if (!this[isFirstChangeKey] && this[cachedValueKey] === value) {
          return;
        }

        const oldValue = this[cachedValueKey];
        this[cachedValueKey] = value;
        const change: SimpleChange<T> = {
          firstChange: this[isFirstChangeKey],
          previousValue: oldValue,
          currentValue: this[cachedValueKey],
          isFirstChange: () => this[isFirstChangeKey],
        };

        callbackFn.call(this, this[cachedValueKey], change);
      },

      get: function () {
        return this[cachedValueKey];
      },
    });
  };
};

export { OnChange };


src/lib/utils/breadth-first-traversal.ts
/**
 * Performs a breadth-first traversal of a tree-like structure
 * and returns the march of the traversal
 *
 * ### Usage
 * ```ts
 * import { breadthFirstTraversal } from '@lerado/typescript-toolbox';
 *
 * const tree = {
 *  id: 1,
 *  children: [
 *    { id: 2, children: [ { id: 4 } ] },
 *    { id: 3, children: [ { id: 5 } ] }
 *  ]
 * };
 *
 * breadthFirstTraversal(tree, 'children', console.log);
 * // => console: 1, 2, 3, 4, 5
 * // => [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }]
 * ```
 * @param root - the tree's root object
 * @param childrenKey - accessor name for child elements
 * @param callback - a callback function to execute for each visited element
 */
const breadthFirstTraversal = <T extends Record<string, unknown>>(
  root: T,
  childrenKey: string,
  callable?: (visitedElement: T) => unknown
): Partial<T>[] => {
  const result: T[] = [];
  const queue: T[] = [];
  queue.push(root);

  while (queue.length) {
    const visitedElement: T = queue.shift() as T;
    // Visit
    if (callable) {
      callable(visitedElement);
    }
    result.push(visitedElement);
    // queue bloc children from right to left to preserve order
    const children: T[] = (visitedElement[childrenKey] as T[]) ?? [];
    for (let childIndex = 0; childIndex < children.length; childIndex++) {
      queue.push(children[childIndex]);
    }
  }

  return result;
};

export { breadthFirstTraversal };


src/lib/utils/strong-password.ts
/**
 * Strong password regular expression
 *
 * **Rules**
 *
 * *- at least eight characters long*
 *
 * *- at least one capital letter*

 * *- at least one small letter*
 *
 * *- at least one number*
 *
 * *- at least one special character*
 */
const STRONG_PASSWORD_REGEXP =
  /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{8,})/;

/**
 * Returns true if a given is considered a strong password, false otherwise
 *
 * ### Usage
 * ```ts
 * import { isStrongPassword } from '@lerado/typescript-toolbox';
 *
 * isStrongPassword('123456');
 * // => false
 * isStrongPassword('password123');
 * // => false
 * isStrongPassword('Password123');
 * // => false
 * isStrongPassword('Password!123');
 * // => true
 * ```
 * @param password - string representing a password to check
 * @see STRONG_PASSWORD_REGEXP
 */
function isStrongPassword(password: string): boolean {
  return STRONG_PASSWORD_REGEXP.test(password);
}

export { STRONG_PASSWORD_REGEXP, isStrongPassword };
